s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py

```diff
@@ -0,0 +1,211 @@
+            src_texts = src_texts,
+            tgt_texts = tgt_texts,
+            speakers = speakers,
+            src_langs = src_langs,
+            tgt_langs = tgt_langs,
+            srs = srs,
+            ids = ids,
+            tgt_dict = tgt_dict,
+            pre_tokenizer = pre_tokenizer,
+            bpe_tokenizer = bpe_tokenizer,
+            upstream_rate = upstream_rate,
+        )
+
+    @classmethod
+    def from_tsv(
+        cls,
+        root: str,
+        data_cfg: S2TDataConfig,
+        splits: str,
+        tgt_dict,
+        pre_tokenizer,
+        bpe_tokenizer,
+        is_train_split: bool,
+        epoch: int,
+        seed: int,
+        upstream_rate: int,
+    ) -> SpeechToTextDataset:
+
+        _splits = splits.split(",")
+        assert len(_splits) == 1, "do not support multiple files training"
+
+        split = _splits[0]
+        tsv_path = op.join(root, f"{split}.tsv")
+        if not op.isfile(tsv_path):
+            raise FileNotFoundError(f"Dataset not found: {tsv_path}")
+        
+        with open(tsv_path) as f:
+            reader = csv.DictReader(
+                f,
+                delimiter="\t",
+                quotechar=None,
+                doublequote=False,
+                lineterminator="\n",
+                quoting=csv.QUOTE_NONE,
+            )
+            samples = [dict(e) for e in reader]
+
+        assert len(samples) > 0
+
+        return cls._from_list(
+            split,
+            is_train_split,
+            [samples],
+            data_cfg,
+            tgt_dict,
+            pre_tokenizer,
+            bpe_tokenizer,
+            upstream_rate,
+        )
+
+
+class S3prl_SpeechToTextDataset(SpeechToTextDataset):
+
+    TARGET_RATE = 16000
+
+    def __init__(self, *args, srs = Optional[List[int]], upstream_rate = 160, **kwargs):
+        
+        super().__init__(*args, **kwargs)
+
+        self.srs = srs
+
+        self.resamplers = {}
+        for sr in set(srs):
+            self.resamplers[sr] = torchaudio.transforms.Resample(
+                orig_freq = sr,
+                new_freq = self.TARGET_RATE,
+            )
+
+        for i in range(len(self.n_frames)):
+            new_n_frames = self.n_frames[i] * self.TARGET_RATE / self.srs[i] / upstream_rate
+            self.n_frames[i] = int(new_n_frames)
+
+    def __getitem__(
+        self, index: int
+    ) -> Tuple[int, torch.Tensor, Optional[torch.Tensor]]:
+
+        source, sr = torchaudio.load(self.audio_paths[index])
+
+        assert self.srs[index] == sr
+        source = self.resamplers[sr](source)
+        source = torch.mean(source, dim=0)
+        source = source.view(-1)
+
+        if self.feature_transforms is not None:
+            assert not self.data_cfg.use_audio_input
+            source = self.feature_transforms(source)
+        source = source.float()
+
+        target = None
+        if self.tgt_texts is not None:
+            tokenized = self.tokenize_text(self.tgt_texts[index])
```

Comment:
it's OK to do it on the fly, but how about making it pre-computed as you load the text? It will save some time from loaders, more or less.

Reply:
I think tokenizing on the fly is a more proper way. For the time consuming issue, the resampling and feature extraction would be the actual bottleneck. And if we have other hidden data to test on, it would be more convenient to take raw text as input.

Reply:
Thanks for the clarification. Make sense to me